internal/characters/tartaglia/abil.go

```diff
@@ -0,0 +1,411 @@
+			}
+			c.rtFlashICD[t.Index()] = c.Core.F + 42 //0.7s icd
+		}, "Riptide Flash", 5)
+	}
+}
+
+func (c *char) rtSlashCallback(t core.Target) {
+	if c.rtExpiry[t.Index()] > c.Core.F {
+		if c.rtSlashICD[t.Index()] > c.Core.F {
+			return
+		}
+
+		c.AddTask(func() {
+			d := c.Snapshot(
+				"Riptide Slash",
+				core.AttackTagElementalArt,
+				core.ICDTagNone,
+				core.ICDGroupDefault,
+				core.StrikeTypeDefault,
+				core.Hydro,
+				25,
+				rtSlash[c.TalentLvlSkill()],
+			)
+			d.Targets = core.TargetAll
+
+			c.Core.Combat.ApplyDamage(&d)
+			c.Core.Log.Debugw("Riptide Slash ticked", "frame", c.Core.F, "event", core.LogCharacterEvent, "dur",
+				c.Core.Status.Duration("tartagliamelee"), "target", t.Index(), "slashICD", c.rtSlashICD[t.Index()], "rtExpiry", c.rtExpiry[t.Index()])
+
+			if c.rtParticleICD < c.Core.F {
+				c.rtParticleICD = c.Core.F + 180 //3 sec
+				c.QueueParticle("tartaglia", 1, core.Hydro, 100)
+			}
+		}, "Riptide Slash", 5)
+		c.rtSlashICD[t.Index()] = c.Core.F + 90 //1.5s icd
+	}
+}
+
+func (c *char) rtBlastCallback(t core.Target) {
+	if c.rtExpiry[t.Index()] > c.Core.F {
+		if c.rtSlashICD[t.Index()] > c.Core.F {
+			return
+		}
+
+		c.AddTask(func() {
+			d := c.Snapshot(
+				"Riptide Blast",
+				core.AttackTagElementalBurst,
+				core.ICDTagNone,
+				core.ICDGroupDefault,
+				core.StrikeTypeDefault,
+				core.Hydro,
+				50,
+				rtBlast[c.TalentLvlBurst()],
+			)
+			d.Targets = core.TargetAll
+
+			c.Core.Combat.ApplyDamage(&d)
+			// triggering riptide blast will clear riptide status
+			c.rtExpiry[t.Index()] = 0
+			c.Core.Log.Debugw("Riptide Blast ticked", "frame", c.Core.F, "event", core.LogCharacterEvent, "dur",
+				c.Core.Status.Duration("tartagliamelee"), "target", t.Index(), "rtExpiry", c.rtExpiry[t.Index()])
+		}, "Riptide Blast", 5)
+	}
+}
+
+//apply riptide status to enemy hit
+func (c *char) applyRT() {
+	c.Core.Events.Subscribe(core.OnDamage, func(args ...interface{}) bool {
+		ds := args[1].(*core.Snapshot)
+		t := args[0].(core.Target)
+		crit := args[3].(bool)
+
+		if ds.AttackTag == core.AttackTagNormal || ds.AttackTag == core.AttackTagExtra {
+			if c.Core.Status.Duration("tartagliamelee") > 0 {
+				if !crit {
+					return false
+				}
+
+				//dont log if it just refreshes riptide status
+				if c.rtExpiry[t.Index()] <= c.Core.F {
+					c.Core.Log.Debugw("Tartaglia applied riptide", "frame", c.Core.F, "event", core.LogCharacterEvent, "target", t.Index(), "rtExpiry", c.Core.F+rtA1)
+				}
+
+				c.rtExpiry[t.Index()] = c.Core.F + rtA1
+			} else {
+				if ds.AttackTag == core.AttackTagNormal {
+					return false
+				}
+				//aim mode
+				//dont log if it just refreshes riptide status
+				if c.rtExpiry[t.Index()] <= c.Core.F {
+					c.Core.Log.Debugw("Tartaglia applied riptide", "frame", c.Core.F, "event", core.LogCharacterEvent, "target", t.Index(), "rtExpiry", c.Core.F+rtA1)
+				}
+
+				c.rtExpiry[t.Index()] = c.Core.F + rtA1
+			}
+		}
```

Comment:
I think the if logic is somewhat confusingly laid out because you need to repeat the "AttackTag" if condition in the ranged section. I think it'd be clearer if you moved the melee vs. ranged logic as the top layer, then the attack tag logic can be the inner layer.